Dockerfile

```diff
@@ -1,29 +1,115 @@
-FROM ruby:2.7.2
-
-# 1. Configuring apt to properly install yarn
-# 2. Installing node, postgres-client, yarn
-# 3. Clearing apt-cached files to tighten the image size
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1 apt-key add - \
-  && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list \
-  && apt-get update -qq \
-  && apt-get install -y build-essential \
-  nodejs \
-  yarn \
-  && rm -rf /var/lib/apt/lists/*
+# Generating a generic Rails image.
+# Just for caching.
+FROM ruby:2.7-alpine AS rails
 
-WORKDIR /app
+# `apk` is like `apt`, but for Alpine Linux.
+RUN apk add --update --no-cache --quiet \
+    build-base \
+    nodejs-current \
+    postgresql-dev \
+    sqlite-dev \
+    tzdata \
+    yarn
+
+# Create a generic Gemfile with only Rails.
+RUN echo -e "source 'https://rubygems.org'\ngem 'rails'" > /Gemfile
+# Install Rails, quietly, in 4 threads.
+RUN bundle install --quiet -j4
```

Comment:
How do we get the correct version of Rails?

Reply:
This is a base image with common rails dependencies.  Basically just a caching layer to ensure that we have most of the gems required in the later images.

The correct version of Rails is installed after copying in the `Gemfile` and `Gemfile.lock`.

Reply:
I don't think is actually helpful, because the gems Rails relies on are all version locked to the Rails version. If we install Rails 6.2.x, we install ActiveRecord 6.2.x, so we'd be reinstalling Rails 6.0.2.3 and ActiveRecord 6.0.2.3. I think this is only helpful if we have the correct Rails version.

Reply:
I also worry this could introduce subtle errors where incompatible versions are used, and I can't diagnose why.

Reply:
This has been removed.  It may require additional installs in later `bundle install` commands, but the optimization is pretty slim, not a big loss.